src/Compilers/Core/Portable/Text/SourceText.cs

```diff
@@ -127,8 +127,8 @@ public static SourceText From(Stream stream, Encoding encoding = null, SourceHas
                 throw new InvalidDataException();
             }
 
-            var checksum = CalculateChecksum(stream, checksumAlgorithm);
-            return new StringText(text, encoding, checksum, checksumAlgorithm);
+            Debug.Assert(encoding != null);
+            return new StringText(text, encoding, checksumAlgorithm: checksumAlgorithm);
```

Comment:
Is it possible that decode/encode cycle will produce different bytes even if we ignore BOM difference? Can this create a problem, for example different source bytes produce the same checksum because of decoding "errors"?